### ISSUE TITLE
hash generated on aws does not correspond with windows

### ISSUE THREAD

**Issue by kianwei96** [grero/DataProcessingTools#70]
Hash value (72a7) generated on AWS for spike train object. Also included branch details in the screenshot.
![image](https://user-images.githubusercontent.com/50561440/91297568-1e415280-e7d1-11ea-8b11-815e96cefa5f.png)

Trying to load the file in windows, it expects a different hash value (9d1b).
![image](https://user-images.githubusercontent.com/50561440/91297781-78daae80-e7d1-11ea-8479-38cf0542e3e5.png)

windows branch details.
![image](https://user-images.githubusercontent.com/50561440/91297857-9740aa00-e7d1-11ea-8930-9bd95929bb34.png)
![image](https://user-images.githubusercontent.com/50561440/91297879-a4f62f80-e7d1-11ea-8192-b1b0af5652c8.png)

Same issue found for other objects (eg. Unity, Umaze, RPLParallel)

**Comment by grero** [comment]
Hm, that's unfortunate. It seems that we cannot use the hash then, unless it stems from the dirs variable, since they would be different on a windows system?

**Comment by kianwei96** [comment]
i added a few lines in the hash function, and tried it on both aws and windows. they both show an empty args, which is to be expected, but somehow the hash value in the file name is different?

![image](https://user-images.githubusercontent.com/50561440/91298815-429e2e80-e7d3-11ea-966d-3cd62657a9dc.png)

**Comment by kianwei96** [comment]
seems like the pickle.dumps is giving different outputs:

b'\x80\x03}q\x00.'
vs
b'\x80\x04}\x94.'

even though self.args are they same

**Comment by grero** [comment]
Hm, OK. I think this means that we'll have to load each file to check the arguments instead of relying on the hash. Or we could try and create our own hash, like what we do for get_shortname for levels.

**Comment by kianwei96** [comment]
i'm not sure what's the issue here though. i was able to work with objects generated on aws, on my windows machine for the past month or so. the issue just cropped in the past few days. from what i can see, the hashing function hasn't been modified recently right?

**Comment by grero** [comment]
That's right, we haven't made any changes to the function. Can you check if the pickle version is the same on both systems?

**Comment by kianwei96** [comment]
print(pickle.format_version) gives 4.0 for both systems

**Comment by kianwei96** [comment]
hash function seems to behave differently for different python versions. makes sense because i just reinstalled anaconda today, and had a version upgrade to 3.8:
![image](https://user-images.githubusercontent.com/50561440/91300594-19cb6880-e7d6-11ea-9c59-61a340e6177b.png)

perhaps we can consider switching to this library (hashlib) instead.
https://stackoverflow.com/questions/39989534/hash-returning-different-values-on-different-oss
https://docs.python.org/3.5/library/hashlib.html?highlight=hashlib

![image](https://user-images.githubusercontent.com/50561440/91300825-7fb7f000-e7d6-11ea-8339-16e2d199f3c2.png)

**Comment by grero** [comment]
I'm starting to think we'll have to create our own hashes. The reason is that, mostly, these hashes are not used for what we are using them for. I don't think we have any guarantees that the hash values they generate won't change with future versions of the library. I'll think a bit this a bit and see if i can come up with something that works.

**Comment by kianwei96** [comment]
I think it should be save to use hashlib with md5 hashing. It’s a standard hash function, which probably has to be platform/version independent. The hashlib library is also packaged in the default python installation I think, so support should be quite reliable

**Comment by grero** [comment]
Another alternative is to keep a database in each directory that maps hashes to dictionaries of arguments. Something like <objname>_index.json, where the keys are the hashes and the values are the arguments that hash represents. When creating an object, the database can be consulted and if a file already exists, it will be loaded. I realise that this adds quite a bit of complexity, so it might not be worth it.

**Comment by kianwei96** [comment]
But yes, I can see how a custom implementation would be more attractive - we can maybe even encode it in such a way that we can see which arguments differ based off the file name alone

**Comment by grero** [comment]
Could be, let's do a check first, though, before we commit to it. Let's find out if hashes computed using hashlib are stable across python versions, and more importantly, across OSs.

**Comment by grero** [comment]
I was experimenting with hashlib initially, but it wasn't clear how you would actually go about hashing data, which is why I opted for the pickle solution. We could have another look at it, though.

**Comment by grero** [comment]
We don't actually need a hash in the traditional sense, just a consistent abbreviation of the arguments used to compute the objects.

**Comment by kianwei96** [comment]
I agree that the database mapping method would be quite a bit more complex. 

Regarding hashlib’s md5 hash, I tested it on Linux and Windows, and on different versions of python on windows, with the same result.

If all the arguments are numeric, we can simply concaternate them together into the suffix of the file name. Haven’t thought of how to handle string arguments, if any, though.

**Comment by grero** [comment]
Sounds promising. Would you like to put together a pull request for an update to the DPObject.hash function using hashlib?

**Comment by kianwei96** [comment]
Sure, will modify and test on aws and windows before submitting a PR

**Comment by grero** [comment]
I think we can just do this for string:
```python
s = "Some string"
hashlib.md5(s.encode('utf-8')).hexdigest()
```

**Comment by grero** [comment]
We'll need to have some backward compatibility so that we can still load the old data.

**Comment by kianwei96** [comment]
For backward compatibility, could we implement a flag for loading objects, such that I’ll load any object in the directory, ignoring the hash value? Not sure how else we can do it

**Comment by grero** [comment]
The easiest fallback could be to simply loop through all computed objects and see if any of them matches the supplied arguments.
